### ISSUE TITLE
'*' matches entire path in fnmatch

### ISSUE THREAD

**Issue by 17ef7e94-fb38-4776-984b-fb2050ef2847** [python/cpython#72904]
BPO | [28718](https://bugs.python.org/issue28718)
--- | :---
Nosy | @serhiy-storchaka, @MojoVampire, @decibel, @Hooloovoo, @tovrstra

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2016-11-16.20:12:07.121>
labels = ['library']
title = "'*' matches entire path in fnmatch"
updated_at = <Date 2019-03-31.12:59:09.371>
user = 'https://github.com/decibel'
```

bugs.python.org fields:
```python
activity = <Date 2019-03-31.12:59:09.371>
actor = 'Toon Verstraelen'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2016-11-16.20:12:07.121>
creator = 'Jim Nasby'
dependencies = []
files = []
hgrepos = []
issue_num = 28718
keywords = []
message_count = 8.0
messages = ['280985', '281017', '281018', '288608', '290624', '307867', '339054', '339256']
nosy_count = 6.0
nosy_names = ['serhiy.storchaka', 'josh.r', 'Jim Nasby', 'Alberto Galera', 'aaron-whitehouse', 'Toon Verstraelen']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue28718'
versions = []
```

</p></details>

### Linked PRs
* gh-106703
* gh-109879

**Comment by 4f09f828-6568-489b-83d9-4f46a6108ef1** [comment]
Note that somebody has forked the standard library to implement this:
https://github.com/kianxineki/python-wildcard
This shows that the actual changes would be pretty small (though pywildcard is based on 2.x code and does not handle the cross-platform slashes you have been discussing).

It is also worth noting that the glob standard library:
https://docs.python.org/3.7/library/glob.html
implements a "recursive" option that has similar behaviour (* does not span path separators whereas ** does) and essentially builds this on top of fnmatch for the actual filename matching. 

I do not think we can change the default behaviour of fnmatch at this point, but I would like to see this behaviour triggered by an optional argument to the various functions, e.g.:
fnmatch.fnmatch(filename, pattern, glob_asterisks=False)
fnmatch.fnmatchcase(filename, pattern, glob_asterisks=False)
fnmatch.filter(names, pattern, glob_asterisks=False)
fnmatch.translate(pattern, glob_asterisks=False)

In each case, if glob_asterisks (or whatever other name we came up with) is true, the behaviour would match the pywildcard behaviour, i.e.:
    **      matches everything
    *       matches in one path level

I look after the glob matching code in duplicity and would like to start using the standard library to do filename matching for us, but we need the above behaviour. I am happy to do the patching if there is a realistic chance of it being accepted.

**Comment by 4f09f828-6568-489b-83d9-4f46a6108ef1** [comment]
Posted to the [Python-ideas] mailing list, as it is proposing a change to a standard library:
https://mail.python.org/pipermail/python-ideas/2017-February/044880.html

Nobody has responded so far, however. I take this as at least no vehement objection to the idea.

**Comment by barneygale** [comment]
I have an implementation of this *for pathlib*:

- #101398

It exploits a simple trick: swapping path separators and newlines, and then matching without setting `re.DOTALL`. This causes `*` (and other wildcards) to not match directory separators.

If folks thought it was a good idea, we could instead put the implementation in an `fnmatch.globmatch()` or `glob.match()` function.

**Comment by barneygale** [comment]
Some other differences between `fnmatch` and `glob`:

- `**/` has a special meaning in `glob.glob()`. You might think it means "any number of characters, followed by one slash". In fact it means "any number of characters, as long as we terminate after a slash". Importantly, this means it can match *nothing at all*, and so a pattern like `/etc/**/**/**/hosts` will match `/etc/hosts`.
- `*` _as a standalone path segment_ also has a special meaning. You might think it means "any number of any non-slash characters". In fact it means "at least one non-slash character", i.e. it always consumes precisely one path component.

**Comment by barneygale** [comment]
PR available: #106703

**Comment by barneygale** [comment]
The PR above adds an argument to `fnmatch.translate()`, but doesn't change `fnmatchcase()`. It think it would better to add a function to the `glob` module:

```python
def match(path: os.PathLike, pattern: os.PathLike) -> bool:
    ...
```

This would take care of supplying `os.path.sep` (and `altsep` if applicable) to `fnmatch.translate(pattern, seps=...)`.

Questions:

- Does this sound OK?
- Should `glob.match()` accept a *recursive* argument?
- Should `glob.match()` accept an *include_hidden* argument?

**Comment by barneygale** [comment]
I've posted about this on discuss.python.org: https://discuss.python.org/t/add-glob-translate-convert-path-with-shell-wildcards-to-regular-expression/31549

**Comment by wimglenn** [comment]
Adding to the [helpful list from tovrstra](https://github.com/python/cpython/issues/72904#issuecomment-1093732137), I can recommend [pathspec](https://github.com/cpburnz/python-pathspec).

**Comment by barneygale** [comment]
Addressed in Python 3.13 / [cf67ebf](https://github.com/python/cpython/commit/cf67ebfb315ce36175f3d425249d7c6560f6d0d5) / #106703 -- we've added a new [`glob.translate()`](https://docs.python.org/3.13/library/glob.html#glob.translate) function that converts a glob expression into a regular expression.

**Comment by hauntsaninja** [comment]
@barneygale minor thing, I noticed that the `pat` argument is documented as `pathname`. Should we make the arg positional-only or rename the arg to match the documentation?
